stake-pool/program/src/processor.rs

```diff
@@ -0,0 +1,598 @@
+            return Err(Error::InvalidProgramAddress.into());
+        }
+        Self::stake_set_stake_auth(
+            accounts,
+            stake_info.key,
+            withdraw_info.key,
+            stake_pool.withdraw_nonce,
+            stake_info.key,
+            staking_info.key,
+        )?;
+        Ok(())
+    }
+
+    /// Processes an [SetOwner](enum.Instruction.html).
+    pub fn process_set_owner(_program_id: &Pubkey, accounts: &[AccountInfo]) -> ProgramResult {
+        let account_info_iter = &mut accounts.iter();
+        let stake_pool_info = next_account_info(account_info_iter)?;
+        let owner_info = next_account_info(account_info_iter)?;
+        let new_owner_info = next_account_info(account_info_iter)?;
+        let new_owner_fee_info = next_account_info(account_info_iter)?;
+
+        let mut stake_pool = State::deserialize(&stake_pool_info.data.borrow())?.stake_pool()?;
+
+        if *owner_info.key != stake_pool.owner {
+            return Err(Error::InvalidInput.into());
+        }
+        if !owner_info.is_signer {
+            return Err(Error::InvalidInput.into());
+        }
+        stake_pool.owner = *new_owner_info.key;
+        stake_pool.owner_fee_account = *new_owner_fee_info.key;
+        State::Init(stake_pool).serialize(&mut stake_pool_info.data.borrow_mut())?;
+        Ok(())
+    }
+    /// Processes an [Instruction](enum.Instruction.html).
+    pub fn process(program_id: &Pubkey, accounts: &[AccountInfo], input: &[u8]) -> ProgramResult {
+        let instruction = StakePoolInstruction::deserialize(input)?;
+        match instruction {
+            StakePoolInstruction::Initialize(init) => {
+                info!("Instruction: Init");
+                Self::process_initialize(program_id, init, accounts)
+            }
+            StakePoolInstruction::Deposit => {
+                info!("Instruction: Deposit");
+                Self::process_deposit(program_id, accounts)
+            }
+            StakePoolInstruction::Withdraw(amount) => {
+                info!("Instruction: Withdraw");
+                Self::process_withdraw(program_id, amount, accounts)
+            }
+            StakePoolInstruction::SetStakingAuthority => {
+                info!("Instruction: SetStakingAuthority");
+                Self::process_set_staking_auth(program_id, accounts)
+            }
+            StakePoolInstruction::SetOwner => {
+                info!("Instruction: SetOwner");
+                Self::process_set_owner(program_id, accounts)
+            }
+        }
+    }
+}
+
+// Test program id for the swap program.
+#[cfg(not(target_arch = "bpf"))]
+const MY_PROGRAM_ID: Pubkey = Pubkey::new_from_array([2u8; 32]);
+
+/// Routes invokes to the token program, used for testing.
+#[cfg(not(target_arch = "bpf"))]
+pub fn invoke_signed<'a>(
+    instruction: &Instruction,
+    account_infos: &[AccountInfo<'a>],
+    signers_seeds: &[&[&[u8]]],
+) -> ProgramResult {
+    let mut new_account_infos = vec![];
```

Comment:
For your own testing, you'll want to add an additional check to make sure that the token or staking program id is passed in, which will let you choose which `process` to call -- see https://github.com/solana-labs/solana-program-library/blob/master/token-swap/program/src/processor.rs#L500